# **Principal Component Analysis**

### **1. Load the Dataset**

In this experiment, I used breast cancer dataset provided by scikit learn library

In [24]:
# Load the important packages
import polars as pl
import numpy as np
import matplotlib.pyplot as plt
import sklearn

In [42]:
# Load dataset from scikit learn
df_cancer = sklearn.datasets.load_breast_cancer(as_frame=True)
# Convert independent features and dependent feature to polars
independent_features = pl.from_pandas(df_cancer["data"])
dependent_feature = pl.DataFrame({"target": df_cancer["target"].values})
# Concat dataframes
df_cancer = pl.concat([independent_features, dependent_feature], how='horizontal')
df_cancer

shape: (569, 31)
┌────────┬─────────┬────────────────┬───────────┬───┬─────────┬──────────┬────────────────┬────────┐
│ mean   ┆ mean    ┆ mean perimeter ┆ mean area ┆ … ┆ worst   ┆ worst    ┆ worst fractal  ┆ target │
│ radius ┆ texture ┆ ---            ┆ ---       ┆   ┆ concave ┆ symmetry ┆ dimension      ┆ ---    │
│ ---    ┆ ---     ┆ f64            ┆ f64       ┆   ┆ points  ┆ ---      ┆ ---            ┆ i64    │
│ f64    ┆ f64     ┆                ┆           ┆   ┆ ---     ┆ f64      ┆ f64            ┆        │
│        ┆         ┆                ┆           ┆   ┆ f64     ┆          ┆                ┆        │
╞════════╪═════════╪════════════════╪═══════════╪═══╪═════════╪══════════╪════════════════╪════════╡
│ 17.99  ┆ 10.38   ┆ 122.8          ┆ 1001.0    ┆ … ┆ 0.2654  ┆ 0.4601   ┆ 0.1189         ┆ 0      │
│ 20.57  ┆ 17.77   ┆ 132.9          ┆ 1326.0    ┆ … ┆ 0.186   ┆ 0.275    ┆ 0.08902        ┆ 0      │
│ 19.69  ┆ 21.25   ┆ 130.0          ┆ 1203.0    ┆ … ┆ 0.243   ┆ 0.3613   ┆ 0.08758        ┆ 0      │
│ 11.42  ┆ 20.38   ┆ 77.58          ┆ 386.1     ┆ … ┆ 0.2575  ┆ 0.6638   ┆ 0.173          ┆ 0      │
│ 20.29  ┆ 14.34   ┆ 135.1          ┆ 1297.0    ┆ … ┆ 0.1625  ┆ 0.2364   ┆ 0.07678        ┆ 0      │
│ …      ┆ …       ┆ …              ┆ …         ┆ … ┆ …       ┆ …        ┆ …              ┆ …      │
│ 21.56  ┆ 22.39   ┆ 142.0          ┆ 1479.0    ┆ … ┆ 0.2216  ┆ 0.206    ┆ 0.07115        ┆ 0      │
│ 20.13  ┆ 28.25   ┆ 131.2          ┆ 1261.0    ┆ … ┆ 0.1628  ┆ 0.2572   ┆ 0.06637        ┆ 0      │
│ 16.6   ┆ 28.08   ┆ 108.3          ┆ 858.1     ┆ … ┆ 0.1418  ┆ 0.2218   ┆ 0.0782         ┆ 0      │
│ 20.6   ┆ 29.33   ┆ 140.1          ┆ 1265.0    ┆ … ┆ 0.265   ┆ 0.4087   ┆ 0.124          ┆ 0      │
│ 7.76   ┆ 24.54   ┆ 47.92          ┆ 181.0     ┆ … ┆ 0.0     ┆ 0.2871   ┆ 0.07039        ┆ 1      │
└────────┴─────────┴────────────────┴───────────┴───┴─────────┴──────────┴────────────────┴────────┘

In the dataset above, we already have the data with 569 rows and 31 columns. We would like to reduce the dimension of this data as low as possible using PCA.

### **2. PCA Processing**

#### **2.1 Calculate Covariance Matrix**

When dealing with PCA, we need to first calculate covariance matrix in order to get eigenvalues and eigenvectors.

In [54]:
independent_features.to_numpy()

array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]])